olp-cpp-sdk-dataservice-read/src/VolatileLayerClientImpl.cpp

```diff
@@ -19,20 +19,89 @@
 
+#include <olp/core/client/CancellationContext.h>
+#include <olp/dataservice/read/PartitionsRequest.h>
+#include "repositories/DataRepository.h"
+#include "repositories/ExecuteOrSchedule.inl"
+#include "repositories/PartitionsRepository.h"
+
 namespace olp {
 namespace dataservice {
 namespace read {
 
+namespace {
+constexpr auto kLogTag = "VolatileLayerClientImpl";
+}
+
 VolatileLayerClientImpl::VolatileLayerClientImpl(
     olp::client::HRN catalog, std::string layer_id,
-    olp::client::OlpClientSettings client_settings) {}
+    olp::client::OlpClientSettings client_settings)
+    : catalog_(std::move(catalog)),
+      layer_id_(std::move(layer_id)),
+      client_settings_(std::move(client_settings)),
+      pending_requests_(new PendingRequests()) {}
 
-VolatileLayerClientImpl::~VolatileLayerClientImpl() = default;
+VolatileLayerClientImpl::~VolatileLayerClientImpl() {
+  pending_requests_->CancelPendingRequests();
+}
 
 olp::client::CancellationToken VolatileLayerClientImpl::GetData(
     DataRequest data_request, Callback callback) {
-  return olp::client::CancellationToken{[]() {}};
-}
+  olp::client::CancellationContext context;
+  auto cancellation_token = olp::client::CancellationToken(
+      [context]() mutable { context.CancelOperation(); });
+
+  int64_t request_key = pending_requests_->GenerateRequestPlaceholder();
+  pending_requests_->Insert(cancellation_token, request_key);
+
+  auto catalog = catalog_;
+  auto client_settings = client_settings_;
+  auto layer_id = layer_id_;
+  auto pending_requests = pending_requests_;
+
+  repository::ExecuteOrSchedule(&client_settings_, [=]() mutable {
+    if (!data_request.GetDataHandle()) {
+      auto response = repository::PartitionsRepository::GetPartitionById(
+          catalog, layer_id, context, data_request, client_settings);
+
+      if (!response.IsSuccessful()) {
+        pending_requests->Remove(request_key);
+
+        if (callback) {
+          callback(response.GetError());
+        }
+        return;
+      }
+
+      const auto& partitions = response.GetResult().GetPartitions();
+      if (partitions.empty()) {
+        pending_requests->Remove(request_key);
+        OLP_SDK_LOG_INFO_F(kLogTag, "Partition %s not found",
+                           data_request.GetPartitionId()
+                               ? data_request.GetPartitionId().get().c_str()
+                               : "<none>");
+        if (callback) {
+          callback(client::ApiError(client::ErrorCode::NotFound,
+                                    "Partition not found"));
+        }
+        return;
+      }
+
+      data_request.WithDataHandle(partitions.front().GetDataHandle());
+    }
+
+    auto blob_response = repository::DataRepository::GetBlobData(
+        catalog_, layer_id_, "volatile-blob", data_request, context,
+        client_settings_);
+
+    pending_requests->Remove(request_key);
+    if (callback) {
+      callback(blob_response);
+    }
```

Comment:
We can override a call to callback and call pending_requests->Remove() one time.
Fetch option Cache with update is not handled.